rust/lance-encoding/src/decoder.rs

```diff
@@ -195,15 +195,396 @@
 
+/// Metadata describing a page in a file
+///
+/// This is typically created by reading the metadata section of a Lance file
+#[derive(Debug)]
+pub struct PageInfo {
+    /// The number of rows in the page
+    pub num_rows: u32,
+    /// The encoding that explains the buffers in the page
+    pub encoding: pb::ArrayEncoding,
+    /// The offsets of the buffers in the file
+    pub buffer_offsets: Arc<Vec<u64>>,
+}
+
+/// Metadata describing a column in a file
+///
+/// This is typically created by reading the metadata section of a Lance file
+pub struct ColumnInfo {
+    /// The metadata for each page in the column
+    pub page_infos: Vec<Arc<PageInfo>>,
+    pub buffer_offsets: Vec<u64>,
+}
+
+impl ColumnInfo {
+    /// Create a new instance
+    pub fn new(page_infos: Vec<Arc<PageInfo>>, buffer_offsets: Vec<u64>) -> Self {
+        Self {
+            page_infos,
+            buffer_offsets,
+        }
+    }
+}
+
+/// The scheduler for decoding batches
+///
+/// Lance decoding is done in two steps, scheduling, and decoding.  The
+/// scheduling tends to be lightweight and should quickly figure what data
+/// is needed from the disk issue the appropriate I/O requests.  A decode task is
+/// created to eventually decode the data (once it is loaded) and scheduling
+/// moves on to scheduling the next page.
+///
+/// Meanwhile, it's expected that a decode stream will be setup to run at the
+/// same time.  Decode tasks take the data that is loaded and turn it into
+/// Arrow arrays.
+///
+/// This approach allows us to keep our I/O parallelism and CPU parallelism
+/// completely separate since those are often two very different values.
+///
+/// Backpressure should be achieved via the I/O service.  Requests that are
+/// issued will pile up if the decode stream is not polling quickly enough.
+/// The [`crate::EncodingsIo::submit_request`] function should return a pending
+/// future once there are too many I/O requests in flight.
+///
+/// TODO: Implement backpressure
+pub struct DecodeBatchScheduler {
+    root_scheduler: SimpleStructScheduler,
+}
+
+impl DecodeBatchScheduler {
+    fn is_primitive(data_type: &DataType) -> bool {
+        match data_type {
+            DataType::Boolean
+            | DataType::Date32
+            | DataType::Date64
+            | DataType::Decimal128(_, _)
+            | DataType::Decimal256(_, _)
+            | DataType::Duration(_)
+            | DataType::Float16
+            | DataType::Float32
+            | DataType::Float64
+            | DataType::Int16
+            | DataType::Int32
+            | DataType::Int64
+            | DataType::Int8
+            | DataType::Interval(_)
+            | DataType::Null
+            | DataType::RunEndEncoded(_, _)
+            | DataType::Time32(_)
+            | DataType::Time64(_)
+            | DataType::Timestamp(_, _)
+            | DataType::UInt16
+            | DataType::UInt32
+            | DataType::UInt64
+            | DataType::UInt8 => true,
```

Comment:
This thinks `FixedSizeList<FixedSizedList<FixedSizeList<int>>>` to be primitive?

Should we reuse https://docs.rs/arrow-schema/51.0.0/arrow_schema/enum.DataType.html#method.is_primitive as well?

i.e.., `

match data_type {
  dt if dt.is_primtive() { true },
  DataType::FixedSizeList(...) => { .. },
}

Reply:
Yes, `FSL<FSL<FSL<int>>>` is primitive.  It's still encoded as just one big buffer of ints.

I think we can reuse `DataType::is_primitive` to simplify this a little since we should be superset of it.  I'll give it a shot.